### PR TITLE
Add early stopping based on loss convergence

### PR DESCRIPTION
**Pull request by kaminow**
The current `EarlyStopping` class works by checking if training is able to improve over the best loss seen, but sometimes this causes training to stop too early when the loss is still varying a lot but there happens to be one epoch with lower loss. This PR adds another early stopping class that checks that stops training if the mean loss from the previous `n` epochs is within some tolerance.

### CODE REVIEW

File: asapdiscovery-ml/asapdiscovery/ml/es.py [hunk 1]
```diff
@@ -66,3 +68,60 @@ def check(self, epoch, loss, wts_dict):
...
+
+        Returns
+        -------
+        bool
+            Whether to stop training
+        """
+        import numpy as np
+
+        # Add most recent loss
+        self.losses += [loss]
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/es.py:113]: Probably need some type checking on `loss` to make sure not appending junk.

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 2]
```diff
@@ -549,9 +582,33 @@ def init(args, rank=False):
...
 
     # Early stopping
-    if args.early_stopping:
-        es = EarlyStopping(args.early_stopping)
-        exp_configure.update({"early_stopping": args.early_stopping})
+    if args.es_type:
+        es_type = args.es_type.lower()
+        if es_type == "best":
+            if args.es_patience <= 0:
+                raise ValueError("Option to --es_patience must be > 0.")
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:589]:
```suggestion
                raise ValueError("Option to --es_patience must be > 0 if  `best` es_type is used")
```

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 3]
```diff
@@ -549,9 +582,33 @@ def init(args, rank=False):
...
+            es = BestEarlyStopping(args.es_patience)
+            exp_configure.update(
+                {
+                    "early_stopping:method": "best",
+                    "early_stopping:patience": args.es_patience,
+                }
+            )
+        elif es_type == "converged":
+            if args.es_n_check <= 0:
+                raise ValueError("Option to --es_n_check must be > 0.")
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:599]:
```suggestion
                raise ValueError("Option to --es_n_check must be > 0 if `converged` es_type is used.")
```

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 4]
```diff
@@ -549,9 +582,33 @@ def init(args, rank=False):
...
+            if args.es_divergence <= 0:
+                raise ValueError("Option to --es_divergence must be > 0.")
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:601]:
```suggestion
                raise ValueError("Option to --es_divergence must be > 0 if `converged` es_type is used.")
```

Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:601]:
```suggestion
                raise ValueError("Option to --es_divergence must be > 0 if `converged` es_type is used.")
```

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small suggestions, but looking good.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small suggestions, but looking good.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small suggestions, but looking good.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small suggestions, but looking good.

**Comment by hmacdope** [comment]
@kaminow apologies for the multiple review, idk exactly how that happened.

**Comment by kaminow** [comment]
haha all good, should be ready for another look now!

**Review by hmacdope — APPROVED** [review]
Looks great @kaminow!

**Pull request merged**